src/python/pants/backend/python/util_rules/pex_from_targets.py

```diff
@@ -354,14 +304,81 @@ async def pex_from_targets(request: PexFromTargetsRequest, python_setup: PythonS
+) -> _RepositoryPex:
+    repository_pex: Pex | None = None
+    if python_setup.requirement_constraints:
+        maybe_constraints_repository_pex = await Get(
+            _RepositoryPex,
+            _ConstraintsRepositoryPexRequest(request),
+        )
+        if maybe_constraints_repository_pex.maybe_pex:
+            repository_pex = maybe_constraints_repository_pex.maybe_pex
+    elif (
+        python_setup.resolve_all_constraints
+        and python_setup.resolve_all_constraints_was_set_explicitly()
+    ):
+        raise ValueError(
+            "`[python].resolve_all_constraints` is enabled, so "
+            "`[python].requirement_constraints` must also be set."
+        )
+    elif request.resolve_and_lockfile:
+        resolve, lockfile = request.resolve_and_lockfile
+        repository_pex = await Get(
+            Pex,
+            PexRequest(
+                description=f"Installing {lockfile} for the resolve `{resolve}`",
+                output_filename=f"{path_safe(resolve)}_lockfile.pex",
+                internal_only=request.internal_only,
+                requirements=Lockfile(
+                    file_path=lockfile,
+                    file_path_description_of_origin=(
+                        f"the resolve `{resolve}` (from "
+                        "`[python].experimental_resolves_to_lockfiles`)"
+                    ),
+                    # TODO(#12314): Hook up lockfile staleness check.
+                    lockfile_hex_digest=None,
+                    req_strings=None,
+                ),
+                interpreter_constraints=request.interpreter_constraints,
+                platforms=request.platforms,
+                additional_args=request.additional_lockfile_args,
+            ),
+        )
+    elif python_setup.lockfile:
+        repository_pex = await Get(
+            Pex,
+            PexRequest(
+                description=f"Installing {python_setup.lockfile}",
+                output_filename="lockfile.pex",
+                internal_only=request.internal_only,
+                requirements=Lockfile(
+                    file_path=python_setup.lockfile,
+                    file_path_description_of_origin=("the option `[python].experimental_lockfile`"),
+                    # TODO(#12314): Hook up lockfile staleness check once multiple lockfiles
+                    # are supported.
+                    lockfile_hex_digest=None,
+                    req_strings=None,
+                ),
+                interpreter_constraints=request.interpreter_constraints,
+                platforms=request.platforms,
+                additional_args=request.additional_lockfile_args,
+            ),
+        )
+    return _RepositoryPex(repository_pex)
+
+
 @rule
 async def _setup_constraints_repository_pex(
-    request: _ConstraintsRepositoryPexRequest, python_setup: PythonSetup
-) -> _ConstraintsRepositoryPex:
+    constraints_request: _ConstraintsRepositoryPexRequest, python_setup: PythonSetup
```

Comment:
This is where there would have been ambiguity, if we'd tried to use _RepositoryPexRequest directly.